python_homework/fastq_filtrator/fastq_filtrator.py

```diff
@@ -0,0 +1,219 @@
+    '''
+    Filter the read by length.
+    '''
+    if read_len < length_bounds[0] or read_len > length_bounds[1]:
+        return False
+
+    return True
+
+
+def filter_quality(read_len, quality_line, quality_threshold):
+    '''
+    Filter the read by quality
+    '''
+    quality = 0
+
+    for character in quality_line:
+        quality += (ord(character) - 33)
+
+    average_quality = quality / read_len
+
+    if average_quality < quality_threshold:
+        return False
+
+    return True
+
+
+def filter_fastq(lines, gc_bounds, length_bounds, quality_threshold):
+    '''
+    Check the quality for each read in fastq file.
+    If passed - return True, else - False.
+    '''
+    seq_line = lines[1][:-2]
+    quality_line = lines[3][:-2]
+    read_len = len(seq_line)
+
+    # filter the read by GC content
+    if gc_bounds != [0, 100]:
+        if not filter_gc(seq_line, read_len, gc_bounds):
+            return False
+
+    # filter the read by length
+    if not filter_length(read_len, length_bounds):
+        return False
+
+    # filter the read by quality
+    if quality_threshold != 0:
+        if not filter_quality(read_len, quality_line, quality_threshold):
+            return False
+
+    # True if the read passed filtration
+    return True
+
+
+def write_filtered_fastq(lines, quality_result, save_filtered, wons, wonf):
+    '''
+    Write (to the end of file) each read to the passed or (if needed) failed file.
+    '''
+    if quality_result:
+        wons.writelines(lines)
+
+    else:
+        if save_filtered:
+            wonf.writelines(lines)
+
+
+def read_fatsq_for_filter(input_fastq, gc_bounds, length_bounds, quality_threshold,
+                          output_file_prefix, save_filtered, read_len=4):
+    """
+    Reads four lines from a fastq file and passes them to filter and write function.
+    """
+    read_from = open(input_fastq, 'r')
+    write_on_success_to = open(output_file_prefix + '_passed.fastq', 'w')
+    write_on_fail_to = None
+
+    if save_filtered:
+        write_on_fail_to = open(output_file_prefix + '_failed.fastq', 'w')
+
+    try:
+        while True:
+
+            try:
+                lines = [next(read_from) for x in range(read_len)]
+                quality_result = filter_fastq(lines, gc_bounds, length_bounds, quality_threshold)
+
+                # Write out fastq file after filter
+                write_filtered_fastq(lines,
+                                     quality_result,
+                                     save_filtered,
+                                     wons=write_on_success_to,
+                                     wonf=write_on_fail_to)
+
+            except StopIteration:
+                return
+
+    finally:
+        read_from.close()
+        write_on_success_to.close()
+
+        if write_on_fail_to:
+            write_on_fail_to.close()
+
+
+def main(input_fastq, output_file_prefix, gc_bounds, length_bounds, quality_threshold, save_filtered):
```

Comment:
Аргументы функции должы были иметь значения по-умолчанию
```suggestion
def main(input_fastq, output_file_prefix, gc_bounds=(0, 100), length_bounds=(0, 2**32), quality_threshold=0, save_filtered=False):
```
Установка дефолтных параметров при парсинге допустима, но это не панацея. Например, если я хотел бы заимпортировать `main`, то был бы лишён возможности вызвать функцию с аргументами по-умолчанию.